feat: server side rendering

mthuret opened the pull request:
This PR add the support of server side rendering through a new API:

`findResults` to be able to get the results from an InstantSearch instance. 
a new kind of `InstantSearch` instance (/server), that can take a "resultsState" props. 

TODO: 

- [x] Improve the SSR guide.
- [x] Check if static getSearchParameters can be removed.
- [x] add the next-app on our example page. 
- [x] `findResults` name could be better. any ideas? (@Haroenv propose `findResultsState`).

Haroenv commented on docgen/src/guide/Server-side_rendering.md:
```diff
@@ -6,8 +6,57 @@ category: guide
 
-React InstantSearch is not yet compatible with server-side rendering. This is one of our priority topics
-right now. We will soon provide API entries to handle this usecase.
+React InstantSearch builds the Algolia query based on every widgets declared. That's why at the first rendering there is no results yet. This is inconvenient when doing server side rendering as the html generated by the server will not contain any hits. 
```
* every widget declared

Haroenv replied:
* there are no results yet

Haroenv commented on docgen/src/guide/Server-side_rendering.md:
```diff
@@ -6,8 +6,57 @@ category: guide
-right now. We will soon provide API entries to handle this usecase.
+React InstantSearch builds the Algolia query based on every widgets declared. That's why at the first rendering there is no results yet. This is inconvenient when doing server side rendering as the html generated by the server will not contain any hits. 
+
+If you want to have a server side rendering application using React InstantSearch you will need to use the special API entries we provide for this particular use-case: 
```
you can use

Haroenv commented on docgen/src/guide/Server-side_rendering.md:
```diff
@@ -6,8 +6,57 @@ category: guide
+If you want to have a server side rendering application using React InstantSearch you will need to use the special API entries we provide for this particular use-case: 
+
+* a new type of `<InstantSearch>` available under `'react-instantsearch/server'`
+* a `findResults` function available under `'react-instantsearch/server'`
```
* is available

Haroenv commented on docgen/src/guide/Server-side_rendering.md:
```diff
@@ -6,8 +6,57 @@ category: guide
+  }
+}
+
+const resultsState = await findResults(App, { searchState });
```
Where is `searchState` defined?

Haroenv commented on docgen/src/guide/Server-side_rendering.md:
````diff
@@ -6,8 +6,57 @@ category: guide
+const resultsState = await findResults(App, { searchState });
+```
+
+Then, once you'll get those results, you can actually give it as a prop to your `<InstantSearch/>` instance. You'll need a mechanism to render your app after fetching those results. 
````
I would add a few more lines here that actually show how it's done in the example.

mthuret replied:
I think in the end, I should remove the code sample. Right now, it's partial so not very useful, and if I go for the full one then:

- it's next.js only
- it's complex
- it will be static and as soon as the next-app demo will update, I'm sure it will be forgotten. 

Maybe that the link to the next.js example is enough. What do you think?

Haroenv replied:
I like this example, maybe all it lacks is a second step where you use the fetched parameters to renderToString. That would be minimal, but still viable

mthuret replied:
I'm not sure to follow here :) The part that is missing is the render of InstantSearch using `resultsState` as prop. 

The need of `renderToString` is hidden under findResults

Haroenv commented on packages/react-instantsearch/scripts/build.sh:
```diff
@@ -23,6 +23,7 @@ babel -q index.js -o dist/index.js &&
 babel -q dom.js -o dist/dom.js &&
 babel -q connectors.js -o dist/connectors.js &&
 babel -q native.js -o dist/native.js
```
you forgot an `&&` here

Haroenv commented on packages/react-instantsearch/src/core/InstantSearch.js:
```diff
@@ -30,6 +30,7 @@ function validateNextProps(props, nextProps) {
  * @propType {func} [onSearchStateChange] - Function to be called everytime a new search is done. Useful for [URL Routing](guide/Routing.html).
  * @propType {object} [searchState] - Object to inject some search state. Switches the InstantSearch component in controlled mode. Useful for [URL Routing](guide/Routing.html).
  * @propType {func} [createURL] - Function to call when creating links, useful for [URL Routing](guide/Routing.html).
+ * @propType {object} [resultsState] - Object or Array to inject some results to be used at first rendering. Useful for [Server Side Rendering](guide/Server-side_rendering.html). This object or array comes from the `findResults` function. 
```
is it an array when multi index, and object when single index? Might want to elaborate a little on that here.

mthuret replied:
Indeed I'm wondering if I should mention this. It's not meant to be used without the usage of "findResults".

Haroenv commented on packages/react-instantsearch/src/core/createConnector.js:
```diff
@@ -52,6 +51,7 @@ export default function createConnector(connectorDesc) {
       static defaultClassNames = Composed.defaultClassNames;
       static propTypes = connectorDesc.propTypes;
       static defaultProps = connectorDesc.defaultProps;
+      static getSearchParameters = connectorDesc.getSearchParameters; //can be removed?
```
Why is it added if it can be removed?

mthuret replied:
Indeed I'm wondering if I should mention this. It's not meant to be used without the usage of "findResults".

Haroenv commented on packages/react-instantsearch/src/core/createConnector.js:
```diff
@@ -65,12 +65,15 @@ export default function createConnector(connectorDesc) {
         const { ais: { store, widgetsManager }, multiIndexContext } = context;
         this.state = {
           props: this.getProvidedProps(props),
+          canRender: false, //indicate if component are rendered on a server or on an client.
```
indicates if a component is rendered with `findResults` (true) or not

Haroenv commented on packages/react-instantsearch/src/core/createConnector.test.js:
```diff
@@ -381,6 +383,54 @@ describe('createConnector', () => {
+        context,
+      });
+
+      expect(onSearchParameters.mock.calls.length).toBe(1);
```
You can use [`.toHaveBeenCalledTimes(1)`](http://facebook.github.io/jest/docs/expect.html#tohavebeencalledtimesnumber)

mthuret replied:
Will do this refactor in another PR (impact a lot of files)

Haroenv replied:
sure, just saw it in this diff, still useful to know 😄

mthuret replied:
Yep sure, way more readable :)

mthuret replied:
At some point, I'll switch to toHaveBeenCalledWith also.

Haroenv commented on packages/react-instantsearch/src/core/createInstantSearchServer.js:
```diff
@@ -0,0 +1,139 @@
+      return acc;
+    }, {});
+
+  let search;
```
can this be refactored to not be reassigned?

Haroenv commented on packages/react-instantsearch/src/core/createInstantSearchServer.test.js:
```diff
@@ -0,0 +1,186 @@
+/* eslint-env jest, jasmine */
+/* eslint-disable no-console */
```
can be removed

harshmaur commented:
@marielaures You can use something like 

`getServerResults` or `findServerResults` or `getResultsFromState` instead of `findResults`

harshmaur commented:
Also let me know, I can improve the docs for you. I want to get this merged quickly. 😅

vvo commented on .babelrc:
```diff
@@ -4,19 +4,33 @@
         "transform-react-jsx-source",
-        "react-hot-loader/babel"
+        "react-hot-loader/babel",
+        "transform-regenerator",
```
What's the requirement for this? (Curious)

mthuret replied:
The usage of await/async in the next app example :)

vvo commented on docgen/src/guide/Server-side_rendering.md:
```diff
@@ -6,8 +6,24 @@ category: guide
 
-React InstantSearch is not yet compatible with server-side rendering. This is one of our priority topics
-right now. We will soon provide API entries to handle this usecase.
+React InstantSearch builds the Algolia query based on every widget declared. That's why at the first rendering there are no results yet. This is inconvenient when doing server side rendering as the html generated by the server will not contain any hits. 
```
I feel like we should start by saying "React InstantSearch is compatible with server-side rendering, here are examples on how to use it:".

Then at the end of the page add a paragraph explaining a bit the inner details like this first sentence.

Otherwise I open this page, I see a sentence that seem negative, I think "well maybe doesn't work, let's move on".

vvo replied:
I would host the app on zeit (or any fancy host), then link it from here "Live demo", "Code".

After this, some paragraph about how it's working and how to implement it without zeit (small example, high level example).

And then some details about why it's done like that (first render undefined, declarative API handles state etc..)

mthuret replied:
Will add the live-demo after the release of SSR.

vvo commented on docgen/src/guide/Server-side_rendering.md:
```diff
@@ -6,8 +6,24 @@ category: guide
+
+Let's see how it works!
+
+First you will need to get the first results using the `findResults` function. This function takes two parameters: 
```
Either show a very small example or directly link to the example, otherwise this level of detail at this place is too much and I am not yet ready to read this/understand this.

vvo commented on packages/react-instantsearch/examples/next-app/.gitignore:
```diff
@@ -0,0 +1,19 @@
+# See https://help.github.com/ignore-files/ for more about ignoring files.
+
+# dependencies
+/node_modules
```
Why start the gitignore with a slash? (curious)

mthuret replied:
Actually I don't know, it was created by the create-next-app.

vvo commented on packages/react-instantsearch/src/core/createInstantSearch.js:
```diff
@@ -25,6 +25,8 @@ export default function createInstantSearch(defaultAlgoliaClient, root) {
       searchState: PropTypes.object,
       onSearchStateChange: PropTypes.func,
+      onSearchParameters: PropTypes.func,
+      resultsState: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
```
Why can it be an object or array?

mthuret replied:
I'm not sure to get it, but it's basically depending if you're using it with mono or multi index.

vvo commented on packages/react-instantsearch/src/core/createInstantSearchServer.js:
```diff
@@ -0,0 +1,136 @@
+import { isEmpty } from 'lodash';
+import cis from './createInstantSearch';
+
+const searchParameters = [];
```
Isn't there any issue having those variables global to the module?

mthuret replied:
As I'm aware of, everything is sequential so shouldn't. I tried it with several use of <InstantSearch/> (use case with several instance in the same page).

vvo replied:
Do you mean everything is synchronous server side wise in all situations of React server side rendering? We just need to ensure searchParameters, indexName and client are write/read in a synchronous mode and that there's no way for them to be erased/replaced at a moment we do not expect like:
- app1 server side render write variable
- app2 server sider render write variables
- app1 server side render tries to read variables
- app1 gets app2 variables
- 🏡 on 🔥

vvo commented on packages/react-instantsearch/src/core/createInstantSearchServer.js:
```diff
@@ -0,0 +1,136 @@
+ */
+
+const createInstantSearch = function(defaultAlgoliaClient, root) {
+  const InstantSearch = cis(defaultAlgoliaClient, root);
```
Let's try to scope client and indexName to this createInstantSearch method. And then forward client and indexName to the other methods when needed.

vvo commented on packages/react-instantsearch/src/core/createInstantSearchServer.js:
```diff
@@ -0,0 +1,136 @@
+};
+
+const findResults = function(App, params) {
+  ReactDom.renderToString(<App {...params} />);
```
Let's try to access client and indexName here:

const element = <App {...params} />;
console.log(element.client/indexName)

mthuret replied:
I'll try to see if it's possible yep.

vvo commented on packages/react-instantsearch/src/core/createInstantSearchManager.js:
```diff
@@ -39,7 +40,7 @@ export default function createInstantSearchManager({
     widgets: initialState,
     metadata: [],
-    results: null,
+    results: resultsState ? resultsState : null,
```
resultsState || null

mthuret commented:
I believe we are almost done => @Haroenv and @vvo if you want to review this one more time :)

Changes since last time: 

- no more global variables under createInstantSearchServer
- better guide
- new server side rendering example (without next.js)
- API entry for `findResultsState`.

mthuret commented on packages/react-instantsearch/src/core/findResultsState.js:
```diff
@@ -0,0 +1,16 @@
+/** 
```
This is the only part I found really annoying. Couldn't succeed to add some js-doc directly under `createInstantSearchServer`. Ended up writing the findResults doc here.

vvo replied:
Legacy js doc module issues, we will have to replace it as soon as it will get too much in our way (its parser is not really comfortable with recent ways of coding).

vvo commented on docgen/src/guide/Server-side_rendering.md:
```diff
@@ -6,8 +6,67 @@ category: guide
+
+We provide two API entries to handle it: 
+
+* a new type of [`<InstantSearch>`](widgets/InstantSearch.html) available under `'react-instantsearch/server'`
```
This is no more the right explanation I believe?

vvo commented on docgen/src/guide/Server-side_rendering.md:
````diff
@@ -6,8 +6,67 @@ category: guide
+
+3. Forward this `resultState` as a prop of your `<InstantSearch>` instance. 
+
+```jsx
````
App change should come first before findResult usage. It must also include the new import statement

vvo commented on docgen/src/guide/Server-side_rendering.md:
````diff
@@ -6,8 +6,67 @@ category: guide
+2. Use [`findResultsState`](server-side-renderings/findResultsState.html) to get a `resultsState` server side. 
+
+```jsx
+server.get('/', async (req, res) => {
````
Must include the import statement to get the App

harshmaur commented:
Yay!!!

vvo commented:
@harshmaur Will be released on Monday July 10th
